lib/clientv2/ippool.go

```diff
@@ -88,6 +97,60 @@ func (r ipPools) Update(ctx context.Context, res *apiv2.IPPool, opts options.Set
 
 // Delete takes name of the IPPool and deletes it. Returns an error if one occurs.
 func (r ipPools) Delete(ctx context.Context, name string, opts options.DeleteOptions) (*apiv2.IPPool, error) {
+	// Deleting a pool requires a little care because of existing endpoints
+	// using IP addresses allocated in the pool.  We do the deletion in
+	// the following steps:
+	// -  disable the pool so no more IPs are assigned from it
+	// -  remove all affinities associated with the pool
+	// -  delete the pool
+	logCxt := log.WithField("Name", name)
+
+	// Get the pool so that we can find the CIDR associated with it.
+	pool, err := r.Get(ctx, name, options.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	logCxt = logCxt.WithField("CIDR", pool.Spec.CIDR)
+
+	// Check the resource version for an update conflict.  We need to do that here since we
+	// have it in hand rather than waiting for the Update or Delete actions to fail.
+	if opts.ResourceVersion != "" && opts.ResourceVersion != pool.ResourceVersion {
+		return nil, cerrors.ErrorResourceUpdateConflict{
+			Identifier: model.ResourceKey{
+				Kind: apiv2.KindIPPool,
+				Name: name,
+			},
+		}
+	}
+
+	// We'll need to release affinities associated with this pool.  We do this even if the
+	// pool was disabled - in which case there *may* be no affine blocks to release.
+	// If the pool is active, set the disabled flag to ensure we stop allocating from this pool.
+	if !pool.Spec.Disabled {
+		logCxt.Debugf("Disabling pool to release affinities")
+		pool.Spec.Disabled = true
+		if _, err := r.Update(ctx, pool, options.SetOptions{}); err != nil {
+			return nil, err
+		}
+
+		// Reset the resource version before the actual delete.
+		opts.ResourceVersion = ""
+	}
+
+	// Convert the CIDR to a network.  This shouldn't fail, but if it does then there
+	// should be no need to release affinities since we can't assigning from this invalid
+	// pool.
+	if _, cidrNet, err := cnet.ParseCIDR(pool.Spec.CIDR); err != nil {
+		logCxt.Debugf("Releasing pool affinities")
+		err = r.client.IPAM().ReleasePoolAffinities(context.Background(), *cidrNet)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// And finally, delete the pool.
+	logCxt.Debugf("Deleting pool")
+	opts.ResourceVersion = ""
```

Comment:
drat - this isn't needed here - since I reset it above only if the pool needs disabling.